python/paddle/distributed/passes/auto_parallel_recompute.py

```diff
@@ -0,0 +1,388 @@
+from paddle.distributed.auto_parallel.utils import naive_set_dist_op_attr_for_program_by_mesh_and_mapping
+
+
+class RecomputeState(ProgramStats):
+    def __init__(self, block, ops):
+        super(RecomputeState, self).__init__(block=block, ops=ops)
+        self._block = block
+        self._ops = ops
+        self.var_op_deps = {}
+
+    def build_stats(self):
+        for i, op in enumerate(self._ops):
+            for name in op.desc.input_arg_names():
+                if name in self.var_op_deps:
+                    self.var_op_deps[name]["var_as_input_ops"].extend([i])
+                else:
+                    self.var_op_deps[name] = {}
+                    self.var_op_deps[name]["var_as_input_ops"] = [i]
+                    self.var_op_deps[name]["var_as_output_ops"] = []
+
+            for name in op.desc.output_arg_names():
+                if name in self.var_op_deps:
+                    self.var_op_deps[name]["var_as_output_ops"].extend([i])
+                else:
+                    self.var_op_deps[name] = {}
+                    self.var_op_deps[name]["var_as_input_ops"] = []
+                    self.var_op_deps[name]["var_as_output_ops"] = [i]
+
+    def get_recompute_segments(self, checkpoints):
+        segments = []
+        start_idx = -1
+        pre_segment_end_idx = -1
+        while start_idx + 1 < len(checkpoints):
+            if start_idx == -1:
+                ckpt_name = checkpoints[start_idx + 1]
+                if ckpt_name not in self.var_op_deps:
+                    start_idx += 1
+                    continue
+                op_idx_list = self.var_op_deps[ckpt_name]["var_as_output_ops"]
+                if op_idx_list:
+                    segments.append([0, max(op_idx_list) + 1])
+            else:
+                flag, min_idx, max_idx = self.is_subgraph(
+                    [checkpoints[start_idx]], [checkpoints[start_idx + 1]])
+                if flag:
+                    min_idx = self._update_segment_start(min_idx,
+                                                         pre_segment_end_idx)
+                    segments.append([min_idx, max_idx + 1])
+                else:
+                    logging.info("Could not recompute op range [{}] - [{}] ".
+                                 format(min_idx, max_idx + 1))
+            start_idx += 1
+
+        for i, (idx1, idx2) in enumerate(segments):
+            logging.info("recompute segment[{}]".format(i))
+            logging.info("segment start op: [{}]: [{}] [{}]".format(self._ops[
+                idx1].desc.type(), self._ops[idx1].desc.input_arg_names(
+                ), self._ops[idx1].desc.output_arg_names()))
+            logging.info("segment end op: [{}]: [{}] [{}]".format(self._ops[
+                idx2 - 1].desc.type(), self._ops[idx2 - 1].desc.input_arg_names(
+                ), self._ops[idx2 - 1].desc.output_arg_names()))
+
+        return segments
+
+    def modify_forward_desc_for_recompute(self, dist_context):
```

Comment:
Please add more comments. For example, what is the purpose of `modify_forward_desc_for_recompute`?

Reply:
done.